keps/sig-storage/1698-generic-ephemeral-volumes/README.md

```diff
@@ -0,0 +1,566 @@
+          mountPath: /var/log
+        - name: scratch
+          mountPath: /scratch
+      volumes:
+      - name: varlog
+        hostPath:
+          path: /var/log
+      - name: scratch
+        ephemeral:
+          metadata:
+            labels:
+              type: fluentd-elasticsearch-volume
+          spec:
+            accessModes: [ "ReadWriteOnce" ]
+            storageClassName: "scratch-storage-class"
+            resources:
+              requests:
+                storage: 1Gi
+```
+
+The DaemonSet controller will create pods with names like
+`fluentd-elasticsearch-b96sd` and the new controller will then add a
+PVC called `fluentd-elasticsearch-b96sd-scratch` for that pod.
+
+## Design Details
+
+### PVC meta data
+
+The namespace is the same as for the pod and the name is the
+concatenation of pod name and the `Volume.Name`. This is guaranteed to
+be unique for the pod and each volume in that pod. Care must be taken
+by the user to not exceed the length limit for object names, otherwise
+volumes cannot be created.
+
+Other meta data is copied from the `VolumeClaimTemplate.ObjectMeta`.
+
+### Preventing accidental collision with existing PVCs
+
+The new controller will only create missing PVCs. It neither needs to
+delete (handled by garbage collection) nor update (`PodSpec.Volumes`
+is immutable) existing PVCs. Therefore there is no risk that the
+new controller will accidentally modify unrelated PVCs.
+
+The volume scheduling library and kubelet must check that a PVC is
+owned by the pod before using it. Otherwise they must ignored it. This
+ensures that a volume is not used accidentally for a pod in case of a
+name conflict.
+
+### Pod events
+
+Errors that occur while creating the PVC will be reported as events
+for the pod and thus be visible to the user. This includes quota
+violations and name collisions.
+
+### Feature gate
+
+The `GenericEphemeralVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `EphemeralVolumeSource`,
+- anything that specifically acts upon an `EphemeralVolumeSource` (scheduler,
+  kubelet, etc.) instead of merely copying it (statefulset controller)
+  accepts the new volume source.
+
+Existing pods with such a volume will not be started when the feature
+gate is off.
+
+### Modifying volumes
+
+Once the PVC for an ephemeral volume has been created, it can be updated
+directly like other PVCs. The new controller will not interfere with that
+because it never updates PVCs. This can be used to control features
+like volume resizing.
+
+### Late binding
+
+Ideally, provisioning should use late binding (aka
+`WaitForFirstConsumer`). The initial implementation assumes that the
+user is taking care of that by selecting that provisioning mode in the
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
```

Comment:
How will we do that? Just ignore the `VolumeBindingMode` in `StorageClass` if specified?

Reply:
Yes.

Reply:
Hmmm... pretty ugly wonder if we can do better. Can discuss during implementation.